lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,398 @@
+        }
+      }
+
+      return null;
+    }
+
+    var matches = subpathVariants.map(processVariant).whereNotNull().toList();
+
+    switch (matches) {
+      case [var path]:
+        return path;
+      case [_, _, ...] && var paths:
+        throw "Unable to determine which of multiple potential resolutions "
+            "found for ${subpath ?? 'root'} in $packageName should be used. "
+            "\n\nFound:\n"
+            "${paths.join('\n')}";
+    }
+    return null;
+  }
+
+  /// Implementation of the `PATTERN_KEY_COMPARE` comparator from
+  /// https://nodejs.org/api/esm.html#resolution-algorithm-specification.
+  int _compareExpansionKeys(String keyA, String keyB) {
+    var baseLengthA = keyA.contains('*') ? keyA.indexOf('*') + 1 : keyA.length;
+    var baseLengthB = keyB.contains('*') ? keyB.indexOf('*') + 1 : keyB.length;
+    if (baseLengthA > baseLengthB) return -1;
+    if (baseLengthB > baseLengthA) return 1;
+    if (!keyA.contains("*")) return 1;
+    if (!keyB.contains("*")) return -1;
+    if (keyA.length > keyB.length) return -1;
+    if (keyB.length > keyA.length) return 1;
+    return 0;
+  }
+
+  /// Returns a file path to `subpath`, resolved in the `exports` object.
+  ///
+  /// Verifies the file exists relative to `packageRoot`. Instances of `*`
+  /// will be replaced with `patternMatch`.
+  ///
+  /// Implementation of `PACKAGE_TARGET_RESOLVE` from the [Resolution Algorithm
+  /// Specification](https://nodejs.org/api/esm.html#resolution-algorithm-specification).
+  String? _packageTargetResolve(
+      String? subpath, Object exports, String packageRoot,
+      [String? patternMatch]) {
+    switch (exports) {
+      case String string:
+        if (!string.startsWith('./')) {
+          throw "Export '$string' must be a path relative to the package root at '$packageRoot'.";
+        }
+        if (patternMatch != null) {
+          var replaced = string.replaceAll(RegExp(r'\*'), patternMatch);
+          var path = p.normalize(p.join(packageRoot, replaced));
+          return fileExists(path) ? path : null;
+        }
+        return p.join(packageRoot, string);
+      case Map<String, dynamic> map:
+        var conditions = ['sass', 'style', 'default'];
+        for (var (key, value) in map.pairs) {
+          if (!conditions.contains(key)) continue;
+          if (_packageTargetResolve(
+                  subpath, value as Object, packageRoot, patternMatch)
+              case var result?) {
+            return result;
+          }
+        }
+        return null;
+
+      case []:
+        return null;
+
+      case List<dynamic> array:
+        for (var value in array) {
+          var result = _packageTargetResolve(
+              subpath, value as Object, packageRoot, patternMatch);
+          if (result != null) {
+            return result;
+          }
+        }
+
+        return null;
+
+      default:
+        throw "Invalid 'exports' value in ${p.join(packageRoot, 'package.json')}";
```

Comment:
Include the value in the error message. Remember, the point of an error message isn't just to tell a user _that_ something has gone wrong, but to help them understand _why_ it's gone wrong so that they can fix it.